### PR TITLE
Allow build systems to specify the files to watch for changes

### LINKED ISSUE #1671: Allow build systems to specify the files that SourceKit-LSP should watch
Build systems managed by `BuildSystemManager` get file change notifications forwarded from the client. But `SourceKitLSPServer` specifies the patterns to watch for. Build systems should be able to specify their file watch patterns and we should re-register the file watching as a new build system is started, which adds new file watch patterns.

### PR DESCRIPTION
**Pull request by ahoppen**
rdar://136014553
Resolves #1671

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Comment by ahoppen** [comment]
@swift-ci Please test macOS

**Review by hamishknight — APPROVED** [review]

File: Sources/BuildSystemIntegration/SwiftPMBuildSystem.swift [hunk 1]
```diff
@@ -249,6 +251,13 @@ package actor SwiftPMBuildSystem: BuiltInBuildSystem {
   ) async throws {
     self.projectRoot = projectRoot
     self.options = options
+    self.fileWatchers =
+      ["Package.swift", "Package.resolved"].map {
+        FileSystemWatcher(globPattern: projectRoot.appending(component: $0).pathString, kind: [.change])
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/BuildSystemIntegration/SwiftPMBuildSystem.swift:256]: That should be fine. It’s unlikely that you keep adding and removing workspaces to a single SourceKit-LSP server instance. You usually have a 1:1 correspondence between SourceKit-LSP server and workspace.

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Comment by ahoppen** [comment]
@swift-ci Please test macOS

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Pull request merged**